### PR TITLE
Use the rust-src tool for fetching rust source

### LINKED ISSUE #833: include_rustc_srcs creates many configured targets
Setting include_rustc_srcs=True creates a configured target for every rust source file. This greatly increases the number of targets in the external dependency.

With include_rustc_srcs=True:
```bash
> bazel query "deps(@rust_darwin_aarch64//...)" | wc
113861  113894 10974007
```

Without:
```bash
> bazel query "deps(@rust_darwin_aarch64//...)" | wc
254     254   15305
```

This has a significant effect on analysis time in my small workspace. Here is a profile of a clean build for a hello world rust binary.

With include_rustc_srcs=True:
```
=== PHASE SUMMARY INFORMATION ===

Total launch phase time         0.042 s    0.84%
Total init phase time           0.007 s    0.15%
Total target pattern evaluation phase time    0.059 s    1.19%
Total interleaved loading-and-analysis phase time    4.052 s   80.99%
Total preparation phase time    0.002 s    0.05%
Total execution phase time      0.839 s   16.77%
Total finish phase time         0.000 s    0.02%
------------------------------------------------
Total run time                  5.003 s  100.00%
```

Without:
```
=== PHASE SUMMARY INFORMATION ===

Total launch phase time         0.042 s    1.55%
Total init phase time           0.005 s    0.22%
Total target pattern evaluation phase time    0.052 s    1.95%
Total interleaved loading-and-analysis phase time    1.838 s   67.94%
Total preparation phase time    0.002 s    0.08%
Total execution phase time      0.764 s   28.24%
Total finish phase time         0.000 s    0.02%
------------------------------------------------
Total run time                  2.706 s  100.00%
```

It appears the filegroup created by include_rustc_srcs=True is only used to support the rust_analyzer rule. Is it possible to point the rust_analyzer rule at the source root without creating all these bazel targets?

### PR DESCRIPTION
**Pull request by csmulhern**
This tool only includes the relevant rust source files, significantly slimming the size of the download, and significantly decreasing the number of bazel targets generated. The concept of a "host tool" is introduced, which is targetless, to accommodate this change.

This partially fixes #833.

### CODE REVIEW

File: rust/private/repository_utils.bzl [hunk 1]
```diff
@@ -463,12 +463,20 @@ def produce_tool_suburl(tool_name, target_triple, version, iso_date = None):
         target_triple: The rust-style target triple of the tool
         version: The version of the tool among "nightly", "beta', or an exact version.
         iso_date: The date of the tool (or None, if the version is a specific version).
-    """
 
+    Returns:
+        str: The fully qualified url path for the specified tool.
+    """
+    url = ""
     if iso_date:
```
Comment (hlopko) [inline rust/private/repository_utils.bzl:471]: There is no excuse for not having a unit test for this function :) I just uploaded https://github.com/bazelbuild/rules_rust/pull/846.

File: rust/private/repository_utils.bzl [hunk 2]
```diff
@@ -463,12 +463,20 @@ def produce_tool_suburl(tool_name, target_triple, version, iso_date = None):
...
-        return "{}/{}-{}-{}".format(iso_date, tool_name, version, target_triple)
-    else:
-        return "{}-{}-{}".format(tool_name, version, target_triple)
+        url += "{}/".format(iso_date)
+    if tool_name:
+        url += tool_name
```
Comment (hlopko) [inline rust/private/repository_utils.bzl:474]: Tool is not an optional argument, let's`fail()` with an error message when None is passed. Same for version.

File: rust/private/repository_utils.bzl (hunk 1 above)
Comment (hlopko) [reply rust/private/repository_utils.bzl:471]: (feel free to merge it and add a test for none in target_triple)

File: rust/private/repository_utils.bzl [hunk 3]
```diff
@@ -463,12 +463,20 @@ def produce_tool_suburl(tool_name, target_triple, version, iso_date = None):
...
+    if version:
+        url += "-{}".format(version)
```
Comment (hlopko) [inline rust/private/repository_utils.bzl:476]: What do you think about alternative code (not tested):

```py
result = "-".join([e for e in [tool_name, target_triple,version] if e])
if iso_date:
  return iso_date + "/" + result
else:
  return result
```  

Combining += and format is imho confusing and cumbersome.

File: rust/private/repository_utils.bzl [hunk 4]
```diff
@@ -477,9 +485,18 @@ def produce_tool_path(tool_name, target_triple, version):
         tool_name: The name of the tool per static.rust-lang.org
         target_triple: The rust-style target triple of the tool
         version: The version of the tool among "nightly", "beta', or an exact version.
-    """
 
-    return "{}-{}-{}".format(tool_name, version, target_triple)
+    Returns:
```
Comment (hlopko) [inline rust/private/repository_utils.bzl:489]: Oh, this is embarassing. This function duplicates most of `produce_tool_suburl`. Would you mind updating `produce_tool_suburl` to call `produce_tool_path`?

File: rust/private/repository_utils.bzl (hunk 2 above)
Comment (csmulhern) [reply rust/private/repository_utils.bzl]: Done.

File: rust/private/repository_utils.bzl (hunk 1 above)
Comment (csmulhern) [reply rust/private/repository_utils.bzl]: Done.

File: rust/private/repository_utils.bzl (hunk 3 above)
Comment (csmulhern) [reply rust/private/repository_utils.bzl]: Done.

File: rust/private/repository_utils.bzl (hunk 4 above)
Comment (csmulhern) [reply rust/private/repository_utils.bzl:481]: Done.

**Review by hlopko — APPROVED** [review]
Looks great, thank you!

**Comment by hlopko** [comment]
CC @UebelAndre FYI.
